File: app/src/main/java/com/se/termproject/ui/customer/MainActivity.java
package com.se.termproject.ui.customer;

import static com.se.termproject.util.ApplicationClass.USER_ID;

import android.annotation.SuppressLint;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.google.android.material.bottomnavigation.BottomNavigationItemView;
import com.google.android.material.navigation.NavigationBarView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.ktx.Firebase;
import com.se.termproject.R;
import com.se.termproject.base.java.BaseActivity;
import com.se.termproject.data.Customer;
import com.se.termproject.data.Review;
import com.se.termproject.databinding.ActivityMainBinding;
import com.se.termproject.ui.customer.home.HomeFragment;
import com.se.termproject.ui.customer.setting.SettingFragment;
import com.se.termproject.ui.customer.history.HistoryFragment;
import com.se.termproject.util.ApplicationClass;
import com.se.termproject.util.SharedPreferencesManagerKt;

public class MainActivity extends BaseActivity<ActivityMainBinding> {
    private static Fragment homeFragment, historyFragment, settingFragment;

    // firebase
    private FirebaseUser user;
    private FirebaseDatabase mDatabase;
    private DatabaseReference mCustomersReference;

    @Override
    protected ActivityMainBinding setViewBinding() {
        return ActivityMainBinding.inflate(getLayoutInflater());
    }

    @Override
    protected void initAfterBinding() {
        user = FirebaseAuth.getInstance().getCurrentUser();

    initReference();
        initData();

        // fragment
        homeFragment = new HomeFragment();
        historyFragment = new HistoryFragment();
        settingFragment = new SettingFragment();

        // initial setting (with HomeFragment)
        replaceFragment(homeFragment);


        initBottomNavigationView();
        initClickListener();
    }

    private void initReference() {
        mDatabase = FirebaseDatabase.getInstance();
        mCustomersReference = mDatabase.getReference("customers");
    }

    private void initData() {
        USER_ID = SharedPreferencesManagerKt.getUserId();
        Customer customer = new Customer(user.getUid(), user.getDisplayName(), user.getEmail(), new Review());
        mCustomersReference.child(USER_ID).addValueEventListener(new ValueEventListener() {

            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.getValue(Customer.class) == null) {
                    mCustomersReference.child(USER_ID).setValue(customer);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) { }
        });
    }

    private void initBottomNavigationView() {
        binding.mainBnv.setOnItemSelectedListener(new NavigationBarView.OnItemSelectedListener() {

            @SuppressLint("NonConstantResourceId")
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                switch(item.getItemId()) {
                    case R.id.main_bnv_history:
                        replaceFragment(historyFragment);
                        return true;
                    case R.id.main_bnv_setting:
                        replaceFragment(settingFragment);
                        return true;
                }
                return false;
            }
        });
    }

    private void initClickListener() {
        binding.mainHomeFab.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                MenuItem home = binding.mainBnv.getMenu().getItem(0);
                home.setChecked(true);
                replaceFragment(homeFragment);
            }
        });
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }
}

File: app/src/main/java/com/se/termproject/ui/shopkeeper/MainActivity.kt
package com.se.termproject.ui.shopkeeper

import android.util.Log
import android.view.MenuItem
import android.widget.ImageView
import android.widget.Toast
import androidx.core.view.GravityCompat
import com.google.android.material.navigation.NavigationView
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.ktx.auth
import com.google.firebase.database.*
import com.google.firebase.ktx.Firebase
import com.se.termproject.R
import com.se.termproject.base.kotlin.BaseActivity
import com.se.termproject.data.Shop
import com.se.termproject.databinding.ActivityShopkeeperBaseBinding
import com.se.termproject.util.getUserId
import com.se.termproject.util.saveUserId

class MainActivity :
    BaseActivity<ActivityShopkeeperBaseBinding>(ActivityShopkeeperBaseBinding::inflate),
    NavigationView.OnNavigationItemSelectedListener {
    companion object {
        private const val TAG = "ACT/MAIN"
        private lateinit var USER_ID: String
    }

    private lateinit var user: FirebaseUser
    private lateinit var mDatabase: FirebaseDatabase
    private lateinit var mShopsReference: DatabaseReference
    private lateinit var mShopReference: DatabaseReference
    private lateinit var mChildEventListener: ChildEventListener

    override fun initAfterBinding() {
        mDatabase = FirebaseDatabase.getInstance()
        mShopsReference = mDatabase.getReference("shops")

        // 사용자 입력 받아오기
        user = Firebase.auth.currentUser!!
        saveUserId(user.uid)
        USER_ID = getUserId()!!

        initData()
        initDrawerLayout()
        initClickListener()
    }

    private fun initData() {
        mShopReference = mDatabase.getReference("shops").child(USER_ID)
        mShopsReference.addValueEventListener(object : ValueEventListener {

            override fun onDataChange(dataSnapshot: DataSnapshot) {
//                if (dataSnapshot.getValue(Shop::class.java) == null) return

                val shop: Shop = dataSnapshot.getValue(Shop::class.java)!!
                Log.d(TAG, "shop: $shop")

                binding.shopkeeperMainLayout.shopkeeperNameTv.text = shop.name
                binding.shopkeeperMainLayout.shopkeeperTotalTableCountTv.text = shop.totalTableCount.toString()
                binding.shopkeeperMainLayout.shopkeeperAvailableTableCountTv.text = shop.availableTableCount.toString()
            }

            override fun onCancelled(databaseError: DatabaseError) { }

        })
    }

    // drawer layout 초기화
    private fun initDrawerLayout() {
        binding.shopkeeperNav.setNavigationItemSelectedListener(this)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            // 등록하기
            R.id.menu_shopkeeper_nav_register_item -> {
                startNextActivity(RegisterActivity::class.java)
            }

            // 수정하기
            R.id.menu_shopkeeper_nav_edit_item -> {
                Toast.makeText(this, "수정하기", Toast.LENGTH_SHORT).show()
            }

            // 이메일 확인
            R.id.menu_shopkeeper_nav_email_item -> {
                Toast.makeText(this, "이메일 확인", Toast.LENGTH_SHORT).show()
            }

            // 이벤트 등록
            R.id.menu_shopkeeper_nav_event_item -> {
                Toast.makeText(this, "이벤트 등록", Toast.LENGTH_SHORT).show()
            }

            // 문의하기
            R.id.menu_shopkeeper_nav_ask_item -> {
                Toast.makeText(this, "문의하기", Toast.LENGTH_SHORT).show()
            }

            // 버전 정보
            R.id.menu_shopkeeper_nav_version_item -> {
                Toast.makeText(this, "버전 정보", Toast.LENGTH_SHORT).show()
            }
        }

        return false
    }

    // click listener 초기화
    private fun initClickListener() {
        binding.shopkeeperMainLayout.shopkeeperMenuSettingIv.setOnClickListener {
            if (!binding.shopkeeperBaseLayout.isDrawerOpen(GravityCompat.END)) {
                binding.shopkeeperBaseLayout.openDrawer(GravityCompat.END)
            }
        }

        val headerView = binding.shopkeeperNav.getHeaderView(0)
        headerView.findViewById<ImageView>(R.id.header_shopkeeper_menu_iv).setOnClickListener {
            binding.shopkeeperBaseLayout.closeDrawer(GravityCompat.END)
        }

        binding.shopkeeperMainLayout.shopkeeperOpenCloseBtn.setOnClickListener {
            Toast.makeText(this, "switch", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }
}
